### PR TITLE
Fix change in env vars in etcd3

### PR DESCRIPTION
**Pull request by joedborg**
Fixes: https://bugs.launchpad.net/charm-etcd/+bug/1845389

Tests are currently failing with etcd 3.3.15

### CODE REVIEW

**Comment by joedborg** [comment]
```
$ ETCDCTL_CACERT=/var/snap/etcd/common/ca.crt ETCDCTL_CERT=/var/snap/etcd/common/client.crt ETCDCTL_KEY=/var/snap/etcd/common/client.key ETCDCTL_API=3 /snap/bin/etcd.etcdctl alarm list
2019-11-08 23:24:07.793870 W | pkg/flags: unrecognized environment variable ETCDCTL_CA_FILE=/var/snap/etcd/common/ca.crt
2019-11-08 23:24:07.794325 W | pkg/flags: unrecognized environment variable ETCDCTL_CERT_FILE=/var/snap/etcd/common/client.crt
2019-11-08 23:24:07.794365 W | pkg/flags: unrecognized environment variable ETCDCTL_KEY_FILE=/var/snap/etcd/common/client.key
{"level":"warn","ts":"2019-11-08T23:24:12.793Z","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-2aee5729-3503-4ac6-b7c9-1aaa2121dafb/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest connection error: connection error: desc = \"transport: authentication handshake failed: x509: certificate is valid for 172.31.12.9, 172.31.29.145, 172.31.33.192, 252.29.145.1, 54.221.135.172, not 127.0.0.1\""}
```

File: README.md [hunk 1]
```diff
@@ -94,6 +96,19 @@ export ETCDCTL_ENDPOINT=https://{ip of etcd host}:2379
...
 ```
 
+Or if you're using etcd 3.x.x:
+
+```shell
+juju expose etcd
+export ETCDCTL_KEY=$(pwd)/client.key
+export ETCDCTL_CERT=$(pwd)/client.crt
+export ETCDCTL_CA_FILE=$(pwd)/ca.crt
+export ETCDCTL_CACERT=https://{ip of etcd host}:2379
```
Comment (Cynerva) [inline README.md]: This doesn't look right

Comment (joedborg) [reply README.md]: Thanks!  Good spot.

File: reactive/etcd.py [hunk 2]
```diff
@@ -518,9 +518,19 @@ def render_default_user_ssl_exports():
...
+        'export ETCDCTL_CERT_FILE={}\n'.format(client_crt),
+        'export ETCDCTL_CA_FILE={}\n'.format(ca_path)
+    ]
+
+    # etcd 3.x.x
+    evars.extend([
+        'export ETCDCTL_KEY={}\n'.format(client_key),
+        'export ETCDCTL_CERT={}\n'.format(client_crt),
+        'export ETCDCTL_CACERT={}\n'.format(ca_path)
+    ])
```
Comment (Cynerva) [inline reactive/etcd.py]: Looks like this handler only runs once, during charm install. You'll need to add something to the upgrade-charm hook handler to support charm upgrades. Hopefully `remove_state('etcd.pillowmints')` is all you need.

**Comment by joedborg** [comment]
Both 3.2 and 3.3 tested.

```
pytest jobs/integration/charm/test_etcd.py --cloud aws/us-east-1 --model default --controller aws-us-east-1

Test session starts (platform: linux, Python 3.7.5, pytest 5.1.2, pytest-sugar 0.9.2)
rootdir: /home/joe/Projects/ck-jenkins, inifile: pytest.ini
plugins: flaky-3.6.1, asyncio-0.10.0, sugar-0.9.2, xdist-1.29.0, forked-1.0.2
collecting ... 
 jobs/integration/charm/test_etcd.py::test_etcd_actions ✓                                                                                                                                            14% █▌        
 jobs/integration/charm/test_etcd.py::test_etcd_scaling ✓                                                                                                                                            29% ██▉       
 jobs/integration/charm/test_etcd.py::test_snapshot_restore s                                                                                                                                        43% ████▍     
 jobs/integration/charm/test_etcd.py::test_leader_status ✓                                                                                                                                           57% █████▊    
 jobs/integration/charm/test_etcd.py::test_config_snapd_refresh ✓                                                                                                                                    71% ███████▎  
 jobs/integration/charm/test_etcd.py::test_cluster_health ✓                                                                                                                                          86% ████████▋ 
 jobs/integration/charm/test_etcd.py::test_leader_knows_all_members ✓                                                                                                                               100% ██████████

Results (563.67s):
       6 passed
       1 skipped
```

**Comment by joedborg** [comment]
[etcd 3.3 support](https://trello.com/c/9NuxNVYw/1414-etcd-33-support)

File: actions/actions.py [hunk 3]
```diff
@@ -33,22 +34,35 @@ def etcdctl_path():
...
+    if endpoints is not False:
+        major, minor, _ = etcdctl_version().split('.')
+        if int(major) >= 3 and int(minor) >= 3:
+            env['ETCDCTL_CACERT'] = opts['ca_certificate_path']
+            env['ETCDCTL_CERT'] = opts['server_certificate_path']
+            env['ETCDCTL_KEY'] = opts['server_key_path']
+            if endpoints is None:
+                leader_address = leader_get('leader_address')
+                endpoints = leader_address
+                etcdctl_cmd += ' --endpoints={}'.format(endpoints)
```
Comment (Cynerva) [inline actions/actions.py:56]: Can we use `http://127.0.0.1:4001` as the endpoint here?

File: reactive/etcd.py [hunk 4]
```diff
@@ -518,9 +518,20 @@ def render_default_user_ssl_exports():
...
+            'export ETCDCTL_KEY={}\n'.format(client_key),
+            'export ETCDCTL_CERT={}\n'.format(client_crt),
+            'export ETCDCTL_CACERT={}\n'.format(ca_path)
+        ]
+    else:
+        evars = [
+            'export ETCDCTL_KEY_FILE={}\n'.format(client_key),
+            'export ETCDCTL_CERT_FILE={}\n'.format(client_crt),
+            'export ETCDCTL_CA_FILE={}\n'.format(ca_path)
+        ]
```
Comment (Cynerva) [inline reactive/etcd.py:535]: This code will need to be re-run if the etcd version changes, i.e. when `config.changed.channel`

**Comment by Cynerva** [comment]
```
jobs/integration/charm/test_etcd.py::test_snapshot_restore s
```

Looks like this test is being skipped. Can you make sure that snapshot and restore work on etcd 3.3? I see etcdctl calls in actions/restore.py (e.g. [here](https://github.com/charmed-kubernetes/layer-etcd/blob/f61ef888c774ec885e037a347a672ba5da99bffb/actions/restore.py#L98)) and it makes me wonder if they need fixing too.

**Comment by joedborg** [comment]
Yeah, I left it alone because they're being manually skipped.  I'm now looking into it, the changes should have fixed compatibility in actions.py, I guess the logic needs fixing in general though.


> Looks like this test is being skipped. Can you make sure that snapshot and restore work on etcd 3.3? I see etcdctl calls in actions/restore.py (e.g. [here](https://github.com/charmed-kubernetes/layer-etcd/blob/f61ef888c774ec885e037a347a672ba5da99bffb/actions/restore.py#L98)) and it makes me wonder if they need fixing too.

File: actions/actions.py (hunk 3 above)
Comment (joedborg) [reply actions/actions.py:56]: Yes, you're right.  I must have been mixing https and 4001 in testing.

**Comment by joedborg** [comment]
I believe the test is being skipped because of a limitation in libjuju:

```
Traceback (most recent call last):
  File "/home/joe/Projects/ck-jenkins/jobs/integration/charm/test_etcd.py", line 105, in test_snapshot_restore
    str(src), str(dst), tools.controller_name, tools.connection
  File "/home/joe/Projects/ck-jenkins/.venv/lib/python3.7/site-packages/juju/unit.py", line 217, in scp_from
    proxy=proxy, scp_opts=scp_opts)
  File "/home/joe/Projects/ck-jenkins/.venv/lib/python3.7/site-packages/juju/machine.py", line 170, in scp_from
    raise NotImplementedError('proxy option is not implemented')
NotImplementedError: proxy option is not implemented
```

**Comment by Cynerva** [comment]
Joe indicated on IRC that he tested the snapshot and restore actions manually. Thanks. We can fix test_snapshot_restore later.

LGTM, but travis is failing due to a drop in code coverage. Do whatever you gotta do to address that and I'm +1 to merging this,

**Review by Cynerva — APPROVED** [review]

**Pull request merged**